xbmc/interfaces/json-rpc/FileOperations.cpp

```diff
@@ -164,21 +165,51 @@ JSONRPC_STATUS CFileOperations::GetDirectory(const std::string &method, ITranspo
 JSONRPC_STATUS CFileOperations::GetFileDetails(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
 {
   std::string file = parameterObject["file"].asString();
-  if (!CFile::Exists(file))
-    return InvalidParams;
+  if (!URIUtils::IsPlugin(file))
+  {
+    if (!CFile::Exists(file))
+      return InvalidParams;
 
-  if (!CFileUtils::RemoteAccessAllowed(file))
-    return InvalidParams;
+    if (!CFileUtils::RemoteAccessAllowed(file))
+      return InvalidParams;
+  }
+  std::string path;
+  CFileItemList items;
+  CFileItemPtr item;
+  
+  if (!URIUtils::IsPlugin(file))
+  {
+    path = URIUtils::GetDirectory(file);
+    if (path.empty() || !CDirectory::GetDirectory(path, items, "", DIR_FLAG_DEFAULTS) || !items.Contains(file))
+      return InvalidParams;
+    item = items.Get(file);
 
-  std::string path = URIUtils::GetDirectory(file);
+    if (!URIUtils::IsUPnP(file))
+      FillFileItem(item, item, parameterObject["media"].asString(), parameterObject);
+  }
+  else
+  {
+    CPluginDirectory plugin;
+    
+    CURL url(file);
+    // Support url like plugin://plugin.name/video-source-name/<file-id>/<command ex.play>
+    // those ending with file.extension/file-id instead of returning object execute playing
+    if (file.back() != '/')
```

Comment:
This tampering with the URL to match the behavior of one plugin is no good, though.

These plugin URLs should only point to a specific **R**esource/media item, not to a "play" command.

Reply:
what if you want to leave abstract layer outside of hardware layer? Like when your plugin server same file from many sources and you want to add those episodes but there is no point on adding 3-4 same files intead you add url to that 'abstract' url that will handle finding the file for you? 

Also its the behavior that author of this feature @AkariDN from https://github.com/xbmc/xbmc/pull/13566 used his test plugin insert `plugin://plugin.video.testlib/tvshows/?sid=12161&eid=3254641` those don't end with with Resource also. 

I will be honest with you Im not a C++ expert, I was most than happy that I did manage to make this work somehow. 

For not this is still fresh feature and I didn't notice any RL plugin example that use it, so I said to myself, hell the most important is it works, there is no guidance how the url should look, right.

Ok, So I shouldn't check if you left command as last parameter and just assume the last part of url will be file ?

Reply:
> 
> 
> This should not be in the JSON-RPC layer.
> 
> It looks like just skipping the `CFile::Exists` check _should_ get the basics working, but the better solution is a new `CPluginFile` from `CFileFactory` to match `CPluginDirectory`, and implement the "existence" check described in #14210.

Sadly no, skipping `CFile::Exists` only make it throw error and not working in the end same as before, https://github.com/xbmc/xbmc/pull/16087/files#diff-cf45f06cbe48e70d575a9b9c707f7dc1R168 you can see that I skipped that by checking if its plugin or not, but was force to use same logic as feature creator to make it work.

Reply:
Oh, I remember now why I did that. I wanted to make it work at cost of not breaking anything that worked. That's why I added that comment that if files ends with file name it would end up playing it, and I didn't want to reinvent this and break anything in my effort to make this work. Adding `/play` or any other command at end of url when adding via python plugin is simple enough to make it before anyone start using this feature and it gave me the ability to make code work with current code base. Also by using `CPluginDirectory::CheckExists` I would have to modify `parameterObject["media"]` types because it expect to get `tvshows, movies, musicvideos` while the normal GetFileDetails expect `video, files... etc` that why I didn't use `CheckExists` to be as much compatible with codebase as possible hoping this would be easy to review code that don't ingerate in more than it should.

Reply:
> the better solution is a new CPluginFile from CFileFactory to match CPluginDirectory, and implement the "existence" check

This "existence" check is designed only for video library scanning/cleaning functionality and should not be used elsewhere. For plugins GetFileDetails should return information from video database (or "Invalid params" if there are no plugin url entries in database) regardless of whether the plugin supports media library scanning or not. You can just play some item in video plugin, after that GetFileDetails should return at least "playcount" info. In short, "GetFileDetails behaviour for plugins" and "plugins as a video source for media library" are different things.

Reply:
Thanks @AkariDN, that makes sense.

And this url stripping is basically doing what `URIUtils::GetDirectory` does, so that's not so bad.